### PR TITLE
Include source ranges for statements, declarations, and parameter lists in the AST dump

### PR DESCRIPTION
**Pull request by bdash**
Source ranges are currently missing from statements, declarations, and parameter lists in the AST dump. This PR addresses that.

For some background: we have a Swift-based domain-specific language that works by analyzing AST dumps and transforming them into new Swift code. In cases where the input doesn't match the constraints of our DSL, we emit errors that point to the problems in the original Swift source. The lack of source range information on these node types makes this difficult for certain types of errors.

### CODE REVIEW

File: include/swift/AST/Stmt.h [hunk 1]
```diff
@@ -128,7 +128,7 @@ class alignas(8) Stmt {
   LLVM_ATTRIBUTE_DEPRECATED(
       void dump() const LLVM_ATTRIBUTE_USED,
       "only for use within the debugger");
-  void print(raw_ostream &OS, unsigned Indent = 0) const;
+  void print(raw_ostream &OS, const ASTContext * = nullptr, unsigned Indent = 0) const;
```
Comment (jrose-apple) [inline include/swift/AST/Stmt.h]: Nitpick: please provide a parameter name here, even if it's just "Ctx".

File: lib/AST/ASTDumper.cpp [hunk 2]
```diff
@@ -1417,20 +1439,28 @@ class PrintStmt : public StmtVisitor<PrintStmt> {
...
+    printCommon(S, "brace_stmt");
+    printASTNodes(S->getElements());
+    PrintWithColorRAII(OS, ParenthesisColor) << ')';
   }
 
-  void printASTNodes(const ArrayRef<ASTNode> &Elements, StringRef Name) {
-    OS.indent(Indent);
-    PrintWithColorRAII(OS, ParenthesisColor) << "(";
-    PrintWithColorRAII(OS, ASTNodeColor) << Name;
+  void printASTNodes(const ArrayRef<ASTNode> &Elements) {
```
Comment (jrose-apple) [inline lib/AST/ASTDumper.cpp]: Why this change?

File: lib/AST/ASTDumper.cpp [hunk 3]
```diff
@@ -424,7 +424,7 @@ namespace {
 
     void printRec(Decl *D) { D->dump(OS, Indent + 2); }
     void printRec(Expr *E) { E->print(OS, Indent + 2); }
-    void printRec(Stmt *S) { S->print(OS, Indent + 2); }
+    void printRec(Stmt *S, const ASTContext& Ctx) { S->print(OS, &Ctx, Indent + 2); }
```
Comment (jrose-apple) [inline lib/AST/ASTDumper.cpp]: Style nitpick: please attach the `&` to the declaration name, not the type.

**Review by jrose-apple — COMMENTED** [review]
I almost want to say "no" on principle—please please don't parse AST dumps! They are not expected to be stable or complete or machine-readable. But this is a good change regardless, so thank you for doing it. I have one review question and two style nitpicks.

File: lib/AST/ASTDumper.cpp (hunk 2 above)
Comment (bdash) [reply lib/AST/ASTDumper.cpp]: This change was to allow `visitBraceStmt` to delegate to `printCommon`, so it can reuse the logic for printing the source range.

**Review by bdash — COMMENTED** [review]
I understand that the AST dump format is unstable and isn't intended to be consumed, but it's the best option we've found for what we're trying to do.

File: lib/AST/ASTDumper.cpp (hunk 2 above)
Comment (jrose-apple) [reply lib/AST/ASTDumper.cpp]: Ah, right, thanks!

**Comment by bdash** [comment]
I've addressed the style issues.

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Review by jrose-apple — APPROVED** [review]

**Comment by jrose-apple** [comment]
Merged. Thanks, Mark!

**Comment by benlangmuir** [comment]
@jrose-apple @bdash  I reverted this because it appears to have broken a test in CI:  https://ci.swift.org/job/oss-swift-incremental-RA-linux-ubuntu-16_04/4126/

Please take a look

**Comment by jrose-apple** [comment]
How did this pass the regular testing? Weird. Thanks, Ben.

**Comment by bdash** [comment]
#16622 was merged after this was tested, and introduced a new `func_decl` check in test/Driver/dump-parse.swift that needs to be updated to accommodate the inclusion of source ranges in the AST dump. I'll take care of that now.

**Comment by bdash** [comment]
#16682 is this PR rebased on master, with the necessary fixes made to tests.
